File: windows/src/GuGuTalk.Core/Services/HotwordStore.cs
using System.Text.Json;
using GuGuTalk.Core.Models;
using Serilog;

namespace GuGuTalk.Core.Services;

public sealed class HotwordStore
{
    private sealed record PersonalLexiconDocument(int Version, IReadOnlyList<PersonalTerm> Terms);

    private static readonly ILogger Logger = Log.ForContext<HotwordStore>();
    private static readonly string StoragePath = Path.Combine(
        Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData),
        "GuGuTalk", "hotwords.json");
    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        PropertyNameCaseInsensitive = true,
        PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
        WriteIndented = true
    };

    private List<PersonalTerm> _terms = [];
    private readonly string _storagePath;

    public IReadOnlyList<PersonalTerm> Terms => _terms;
    public IReadOnlyList<TextReplacement> Replacements => _terms
        .SelectMany(term => term.Aliases.Select(alias => new TextReplacement(alias, term.Text)))
        .ToArray();
    public bool IsEmpty => _terms.Count == 0;

    public HotwordStore() : this(StoragePath) { }

    internal HotwordStore(string storagePath)
    {
        _storagePath = storagePath;
        Load();
    }

    public void AddTerm(string text)
    {
        string value = text.Trim();
        if (string.IsNullOrEmpty(value)) return;
        if (_terms.Any(term => string.Equals(term.Text, value, StringComparison.OrdinalIgnoreCase))) return;
        _terms.Add(new PersonalTerm(value));
        SortAndSave();
    }

    public void RemoveTerm(Guid id)
    {
        _terms.RemoveAll(term => term.Id == id);
        Save();
    }

    // Retained for migration-compatible callers and explicit aliases from older builds.
    public void Add(string from, string to)
    {
        string alias = from.Trim();
        string canonical = to.Trim();
        if (string.IsNullOrEmpty(alias) || string.IsNullOrEmpty(canonical)) return;

        int index = _terms.FindIndex(term =>
            string.Equals(term.Text, canonical, StringComparison.OrdinalIgnoreCase));
        if (index >= 0)
        {
            if (_terms[index].Aliases.Any(value =>
                    string.Equals(value, alias, StringComparison.OrdinalIgnoreCase))) return;
            _terms[index] = _terms[index] with { Aliases = [.. _terms[index].Aliases, alias] };
        }
        else
        {
            _terms.Add(new PersonalTerm(Guid.NewGuid(), canonical, [alias]));
        }
        SortAndSave();
    }

    public void Remove(string from)
    {
        _terms = _terms.Select(term => term with
        {
            Aliases = term.Aliases.Where(alias => alias != from).ToArray()
        }).ToList();
        Save();
    }

    public string ApplyReplacements(string text)
    {
        string result = text;
        var matches = _terms.SelectMany(term => term.Aliases
                .Append(term.Text)
                .Select(pattern => (Pattern: pattern, Replacement: term.Text)))
            .Where(match => !string.IsNullOrEmpty(match.Pattern))
            .OrderByDescending(match => match.Pattern.Length);

        foreach (var match in matches)
        {
            result = result.Replace(
                match.Pattern,
                match.Replacement,
                StringComparison.OrdinalIgnoreCase);
        }
        return result;
    }

    private void Load()
    {
        try
        {
            if (!File.Exists(_storagePath)) return;
            string json = File.ReadAllText(_storagePath);

            if (json.TrimStart().StartsWith('{'))
            {
                var document = JsonSerializer.Deserialize<PersonalLexiconDocument>(json, JsonOptions);
                if (document?.Terms is not null)
                {
                    _terms = Normalize(document.Terms);
                    return;
                }
            }

            var legacy = JsonSerializer.Deserialize<List<TextReplacement>>(json, JsonOptions) ?? [];
            foreach (var replacement in legacy)
            {
                AddInMemory(replacement.Pattern, replacement.Replacement);
            }
            _terms = Normalize(_terms);
            Save();
        }
        catch (Exception ex)
        {
            Logger.Warning(ex, "Failed to load personal lexicon");
            _terms = [];
        }
    }

    private void AddInMemory(string from, string to)
    {
        string alias = from.Trim();
        string canonical = to.Trim();
        if (string.IsNullOrEmpty(alias) || string.IsNullOrEmpty(canonical)) return;
        int index = _terms.FindIndex(term =>
            string.Equals(term.Text, canonical, StringComparison.OrdinalIgnoreCase));
        if (index >= 0)
        {
            _terms[index] = _terms[index] with { Aliases = [.. _terms[index].Aliases, alias] };
        }
        else
        {
            _terms.Add(new PersonalTerm(Guid.NewGuid(), canonical, [alias]));
        }
    }

    private static List<PersonalTerm> Normalize(IEnumerable<PersonalTerm> input)
    {
        var result = new List<PersonalTerm>();
        foreach (var term in input)
        {
            string text = term.Text.Trim();
            if (string.IsNullOrEmpty(text)) continue;
            var aliases = term.Aliases
                .Select(value => value.Trim())
                .Where(value => !string.IsNullOrEmpty(value))
                .Distinct(StringComparer.OrdinalIgnoreCase)
                .OrderBy(value => value, StringComparer.OrdinalIgnoreCase)
                .ToArray();
            int existing = result.FindIndex(value =>
                string.Equals(value.Text, text, StringComparison.OrdinalIgnoreCase));
            if (existing >= 0)
            {
                result[existing] = result[existing] with
                {
                    Aliases = result[existing].Aliases.Concat(aliases)
                        .Distinct(StringComparer.OrdinalIgnoreCase)
                        .OrderBy(value => value, StringComparer.OrdinalIgnoreCase)
                        .ToArray()
                };
            }
            else
            {
                result.Add(new PersonalTerm(term.Id == Guid.Empty ? Guid.NewGuid() : term.Id, text, aliases));
            }
        }
        return result.OrderBy(term => term.Text, StringComparer.OrdinalIgnoreCase).ToList();
    }

    private void SortAndSave()
    {
        _terms = _terms.OrderBy(term => term.Text, StringComparer.OrdinalIgnoreCase).ToList();
        Save();
    }

    private void Save()
    {
        try
        {
            Directory.CreateDirectory(Path.GetDirectoryName(_storagePath)!);
            string json = JsonSerializer.Serialize(new PersonalLexiconDocument(2, _terms), JsonOptions);
            File.WriteAllText(_storagePath, json);
        }
        catch (Exception ex)
        {
            Logger.Warning(ex, "Failed to save personal lexicon");
        }
    }
}

File: macos/Sources/DesktopVoiceInput/Models/HotwordStore.swift
import Foundation

struct TextReplacement: Codable, Identifiable, Equatable, Sendable {
    var id: String { from }
    var from: String
    var to: String
}

struct PersonalTerm: Codable, Identifiable, Equatable, Sendable {
    let id: UUID
    var text: String
    var aliases: [String]

    init(id: UUID = UUID(), text: String, aliases: [String] = []) {
        self.id = id
        self.text = text
        self.aliases = aliases
    }
}

private struct PersonalLexiconDocument: Codable {
    let version: Int
    let terms: [PersonalTerm]
}

@MainActor
final class HotwordStore: ObservableObject {
    @Published private(set) var terms: [PersonalTerm] = []

    private static let storageKey = "personalLexicon"
    private static let legacyStorageKey = "textReplacements"
    private let defaults: UserDefaults

    init(defaults: UserDefaults = .standard) {
        self.defaults = defaults
        load()
    }

    var isEmpty: Bool { terms.isEmpty }

    var replacements: [TextReplacement] {
        terms.flatMap { term in
            term.aliases.map { TextReplacement(from: $0, to: term.text) }
        }
    }

    func addTerm(_ text: String) {
        let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return }
        guard !terms.contains(where: { $0.text.caseInsensitiveCompare(trimmed) == .orderedSame }) else {
            return
        }

        terms.append(PersonalTerm(text: trimmed))
        sortAndSave()
    }

    func removeTerm(id: UUID) {
        terms.removeAll { $0.id == id }
        save()
    }

    // Kept for migration-compatible callers and exact aliases added by older builds.
    func add(from: String, to: String) {
        let trimmedFrom = from.trimmingCharacters(in: .whitespacesAndNewlines)
        let trimmedTo = to.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmedFrom.isEmpty, !trimmedTo.isEmpty else { return }

        if let index = terms.firstIndex(where: { $0.text.caseInsensitiveCompare(trimmedTo) == .orderedSame }) {
            guard !terms[index].aliases.contains(where: {
                $0.caseInsensitiveCompare(trimmedFrom) == .orderedSame
            }) else { return }
            terms[index].aliases.append(trimmedFrom)
        } else {
            terms.append(PersonalTerm(text: trimmedTo, aliases: [trimmedFrom]))
        }
        sortAndSave()
    }

    func remove(_ from: String) {
        for index in terms.indices {
            terms[index].aliases.removeAll { $0 == from }
        }
        terms.removeAll { $0.aliases.isEmpty && $0.text == from }
        save()
    }

    func applyReplacements(to text: String) -> String {
        var result = text
        let matches = terms.flatMap { term in
            ([term.text] + term.aliases).map { (pattern: $0, replacement: term.text) }
        }.sorted { $0.pattern.count > $1.pattern.count }

        for match in matches where !match.pattern.isEmpty {
            result = result.replacingOccurrences(
                of: match.pattern,
                with: match.replacement,
                options: .caseInsensitive
            )
        }
        return result
    }

    private func load() {
        if let data = defaults.data(forKey: Self.storageKey) {
            if let document = try? JSONDecoder().decode(PersonalLexiconDocument.self, from: data) {
                terms = normalizedTerms(document.terms)
                return
            }
            if let decoded = try? JSONDecoder().decode([PersonalTerm].self, from: data) {
                terms = normalizedTerms(decoded)
                return
            }
        }

        guard let legacyData = defaults.data(forKey: Self.legacyStorageKey),
              let legacy = try? JSONDecoder().decode([TextReplacement].self, from: legacyData) else {
            return
        }

        var migrated: [PersonalTerm] = []
        for replacement in legacy {
            let canonical = replacement.to.trimmingCharacters(in: .whitespacesAndNewlines)
            let alias = replacement.from.trimmingCharacters(in: .whitespacesAndNewlines)
            guard !canonical.isEmpty, !alias.isEmpty else { continue }

            if let index = migrated.firstIndex(where: {
                $0.text.caseInsensitiveCompare(canonical) == .orderedSame
            }) {
                if !migrated[index].aliases.contains(where: {
                    $0.caseInsensitiveCompare(alias) == .orderedSame
                }) {
                    migrated[index].aliases.append(alias)
                }
            } else {
                migrated.append(PersonalTerm(text: canonical, aliases: [alias]))
            }
        }

        terms = normalizedTerms(migrated)
        save()
    }

    private func normalizedTerms(_ input: [PersonalTerm]) -> [PersonalTerm] {
        var result: [PersonalTerm] = []
        for var term in input {
            term.text = term.text.trimmingCharacters(in: .whitespacesAndNewlines)
            guard !term.text.isEmpty else { continue }
            term.aliases = Array(Set(term.aliases.map {
                $0.trimmingCharacters(in: .whitespacesAndNewlines)
            }.filter { !$0.isEmpty })).sorted()

            if let index = result.firstIndex(where: {
                $0.text.caseInsensitiveCompare(term.text) == .orderedSame
            }) {
                result[index].aliases = Array(Set(result[index].aliases + term.aliases)).sorted()
            } else {
                result.append(term)
            }
        }
        return result.sorted { $0.text.localizedCompare($1.text) == .orderedAscending }
    }

    private func sortAndSave() {
        terms.sort { $0.text.localizedCompare($1.text) == .orderedAscending }
        save()
    }

    private func save() {
        let document = PersonalLexiconDocument(version: 2, terms: terms)
        guard let data = try? JSONEncoder().encode(document) else { return }
        defaults.set(data, forKey: Self.storageKey)
    }
}
